### DISCUSSION TITLE
`@remix-run/node` confusion

### DISCUSSION THREAD

**Discussion by Download** [remix-run/remix#4650] [General]
I am starting with Remix and confused about `@remix-run/node` a bit. This confusion started when I was reading the [data loading docs](https://remix.run/docs/en/v1/guides/data-loading). There in [section Basics](https://remix.run/docs/en/v1/guides/data-loading#basics) it shows:

```typescript
import type { LoaderFunction } from "@remix-run/node"; // or cloudflare/deno
import { json } from "@remix-run/node"; // or cloudflare/deno
import { useLoaderData } from "@remix-run/react";

export const loader: LoaderFunction = async () => {
  return json([
    { id: "1", name: "Pants" },
    { id: "2", name: "Jacket" },
  ]);
};

// ...
```

I find it interesting to see the type `LoaderFunction` and the utility function `json` coming from `@remix-run/node` here. So I'm digging into the [`@remix-run/node` docs](https://remix.run/docs/en/v1/other-api/node). And this is where the confusion starts. This page starts out by saying:

> Most of the time you will never be importing from this package directly as it's used internally by adapters such as @remix-run/express.

And then documents only a bit about polyfills and the `installGlobals` function. I cannot find the docs for the `LoaderFunction` type and the `json` helper function.

This leaves me confused. Is the basic example of a loader function showing us an exceptional way of writing loaders? Especially the type `LoaderFunction` seems very generic and something we would want to use often. So does the signature of `LoaderFunction` actually change when on different platforms? And where can I find the docs for this type? Should I actually be writing my loader functions in a different way to remain independent of the selected platform for deployment?

Any insights would be appreciated.

**Comment by lili21** [comment]
it wont change on different platform.

**Comment by davidhousedev** [comment]
I was confused about the same thing. Simply put, the example you linked demonstrates the way that loaders should be written.

Have you read at all into the handling for Adapters? These are functions that transform the Request/Params data from specific server environments and cloud environments into the Web Fetch API types represented in `LoaderFunction`. Adapters handle portability, so you can write portable loaders. Imports from the `remix/express` package create an indirection layer between Express and Remix, where you will use types imported from `remix/node`.

I'm still learning Remix, so others might have a more clarity to add here!
